Backport from Mandrake
(fixes kernel addresses leak through error code)

--- linux-2.4.19/arch/i386/mm/fault.c.leak	2004-02-06 17:25:37.000000000 -0700
+++ linux-2.4.19/arch/i386/mm/fault.c	2004-02-06 17:26:55.000000000 -0700
@@ -271,7 +271,8 @@
 	/* User mode accesses just cause a SIGSEGV */
 	if (error_code & 4) {
 		tsk->thread.cr2 = address;
-		tsk->thread.error_code = error_code;
+		/* Kernel addresses are always protection faults */
+		tsk->thread.error_code = error_code | (address >= TASK_SIZE);
 		tsk->thread.trap_no = 14;
 		info.si_signo = SIGSEGV;
 		info.si_errno = 0;
